OpenRA.Game/Activities/Activity.cs

```diff
@@ -9,50 +9,216 @@
  */
 #endregion
 
+using System;
 using System.Collections.Generic;
 using System.Linq;
 using OpenRA.Traits;
 
 namespace OpenRA.Activities
 {
+	public enum ActivityState { Queued, Active, Done, Canceled }
+
 	public abstract class Activity
 	{
-		public Activity NextActivity { get; set; }
+		public ActivityState State { get; private set; }
+
+		public Activity Root
+		{
+			get
+			{
+				if (Parent == null)
+					return this;
+
+				var p = Parent;
+				while (p.Parent != null)
+					p = p.Parent;
+
+				return p;
+			}
+		}
+
+		private Activity parent;
+		public Activity Parent
+		{
+			get
+			{
+				return parent;
+			}
+
+			protected set
+			{
+				parent = value;
+
+				var next = nextActivity;
+				if (next != null)
+					next.Parent = parent;
+			}
+		}
+
+		private Activity childActivity;
+		protected Activity ChildActivity
+		{
+			get
+			{
+				return childActivity != null && childActivity.State < ActivityState.Done ? childActivity : null;
+			}
+
+			set
+			{
+				if (value == this || value == Parent || value == nextActivity)
+					childActivity = null;
+				else
+				{
+					childActivity = value;
+
+					if (childActivity != null)
+						childActivity.Parent = this;
+				}
+			}
+		}
+
+		protected Activity nextActivity;
+		public virtual Activity NextActivity
+		{
+			get
+			{
+				return nextActivity != null ? nextActivity : Parent;
+			}
+
+			set
+			{
+				if (value == this || value == Parent || (value != null && value.Parent == this))
+					nextActivity = null;
+				else
+				{
+					nextActivity = value;
+
+					if (nextActivity != null)
+						nextActivity.Parent = Parent;
+				}
+			}
+		}
+
 		public bool IsInterruptible { get; protected set; }
-		protected bool IsCanceled { get; private set; }
+		public bool IsCanceled { get { return State == ActivityState.Canceled; } }
 
 		public Activity()
 		{
 			IsInterruptible = true;
 		}
 
+		public Activity RealTick(Actor self)
+		{
+			if (State == ActivityState.Done && Game.Settings.Debug.StrictActivityChecking)
+				throw new InvalidOperationException("Actor {0} attempted to tick activity {1} after it had already completed.".F(self, this.GetType()));
+
+			if (State == ActivityState.Queued)
+			{
+				OnFirstRun(self);
+				State = ActivityState.Active;
+			}
+
+			var ret = Tick(self);
+			if (ret == null || (ret != this && ret.Parent != this))
```

Comment:
I can't think of a situation where levels could be skipped like that. The change in `ActivityUtils` should prevent it.